add pdctl interactive mode autocomplete

jiashiwen opened the pull request:
### What problem does this PR solve?
#2299

### What is changed and how it works?
interactive mode ,use "tab" to auto complete command and flags

### Check List
Tests


- Manual test (add detailed scripts or steps below)

run pd-ctl as interactive mode
press 'tab' show command will be used

### Code changes

add new varable
```
readlinecompleter *readline.PrefixCompleter
```

add new func
```
func GenCompleter(cmd *cobra.Command) []readline.PrefixCompleterInterface {
	pc := []readline.PrefixCompleterInterface{}
	if len(cmd.Commands()) != 0 {
		for _, v := range cmd.Commands() {
			// fmt.Println(strings.Split(v.Use, " ")[0]
			if v.HasFlags() {
				flagspc := []readline.PrefixCompleterInterface{}

				flaguseages := strings.Split(strings.Trim(v.Flags().FlagUsages(), " "), "\n")

				for i := 0; i < len(flaguseages)-1; i++ {
					flagspc = append(flagspc, readline.PcItem(strings.Split(strings.Trim(flaguseages[i], " "), " ")[0]))
				}
				flagspc = append(flagspc, GenCompleter(v)...)
				pc = append(pc, readline.PcItem(strings.Split(v.Use, " ")[0], flagspc...))

			} else {
				pc = append(pc, readline.PcItem(strings.Split(v.Use, " ")[0], GenCompleter(v)...))
			}
		}
	}
	return pc
}
```

func getMainCmd add new expression
```
readlinecompleter = readline.NewPrefixCompleter(GenCompleter(rootCmd)...)
```

func loop() add
```
AutoComplete: readlinecompleter,
```


### Release note

JmPotato commented:
/rebuild

Yisaer commented on tools/pd-ctl/pdctl/ctl_test.go:
```diff
@@ -0,0 +1,73 @@
+package pdctl
```
Please add the license

jiashiwen replied:
what license?

jiashiwen replied:
// Copyright 2016 TiKV Project Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// See the License for the specific language governing permissions and
// limitations under the License.

?

Yisaer commented on tools/pd-ctl/pdctl/ctl_test.go:
```diff
@@ -0,0 +1,73 @@
+package pdctl
+
+import (
+	"github.com/spf13/cobra"
```
please format the goimports.

jiashiwen replied:
i use idea auto format，what format need？

jiashiwen replied:
the format like this?
import (
	"testing"

	"github.com/spf13/cobra"
)

Yisaer commented on tools/pd-ctl/pdctl/ctl_test.go:
```diff
@@ -0,0 +1,73 @@
+	"testing"
+)
+
+func newACommand() *cobra.Command {
+	a := &cobra.Command{
+		Use:   "testa",
+		Short: "test A command",
+	}
+	return a
+}
+
+func newBCommand() *cobra.Command {
+	a := &cobra.Command{
+		Use:   "testb",
+		Short: "test b command",
+	}
+	return a
+}
+func newDEFCommand() *cobra.Command {
+	a := &cobra.Command{
+		Use:   "testdef",
+		Short: "test def command",
+	}
+	return a
+}
+
+func newCCommand() *cobra.Command {
+	a := &cobra.Command{
+		Use:   "testc",
+		Short: "test c command",
+	}
+	return a
+}
```
Can we simplify these methods into one method which receive the parameters of `Use` and `Short`?

jiashiwen replied:
recieve use

jiashiwen replied:
I just simulate the use of cobra in pdctl

JmPotato commented on tools/pd-ctl/pdctl/ctl_test.go:
```diff
@@ -0,0 +1,87 @@
+// Copyright 2016 TiKV Project Authors.
```
```suggestion
// Copyright 2020 TiKV Project Authors.
```

rleungx commented:
The PR description is not very clear, would you like to format it?

jiashiwen commented:
> description

how to do?

Yisaer commented on tools/pd-ctl/pdctl/ctl_test.go:
```diff
@@ -0,0 +1,87 @@
+	"github.com/spf13/cobra"
+)
+
+func newACommand() *cobra.Command {
+	a := &cobra.Command{
+		Use:   "testa",
+		Short: "test A command",
+	}
+	return a
+}
+
+func newBCommand() *cobra.Command {
+	a := &cobra.Command{
+		Use:   "testb",
+		Short: "test b command",
+	}
+	return a
+}
+func newDEFCommand() *cobra.Command {
+	a := &cobra.Command{
+		Use:   "testdef",
+		Short: "test def command",
+	}
+	return a
+}
+
+func newCCommand() *cobra.Command {
+	a := &cobra.Command{
+		Use:   "testc",
+		Short: "test c command",
+	}
+	return a
+}
```
I think these code can be simplified as below:

```golang
func newCommand(usage, short string) *cobra.Command {
	a := &cobra.Command{
		Use:   usage,
		Short: short,
	}
	return a
}
```

jiashiwen replied:
got it

Yisaer reviewed (changes requested):
The rest LGTM

rleungx commented on tools/pd-ctl/pdctl/ctl.go:
```diff
@@ -37,9 +38,10 @@ type CommandFlags struct {
+	detach            bool
+	interact          bool
+	version           bool
+	readlinecompleter *readline.PrefixCompleter
```
According to go naming convention https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps, it's better to use `readlineCompleter` instead of `readlinecompleter`. And the rest is the same.

jiashiwen replied:
got it

jiashiwen commented:
idc-jenkins-ci/test error,but the unit test not my code

JmPotato commented:
/rebuild

JmPotato commented:
/rebuild

Yisaer reviewed (approved):
LGTM

howardlau1999 commented on tools/pd-ctl/pdctl/ctl.go:
```diff
@@ -190,3 +194,23 @@ func loop() {
+			if v.HasFlags() {
+				flagspc := []readline.PrefixCompleterInterface{}
+				flaguseages := strings.Split(strings.Trim(v.Flags().FlagUsages(), " "), "\n")
+				for i := 0; i < len(flaguseages)-1; i++ {
```
`flaguseages` -> `flagUsages`

rleungx commented on tools/pd-ctl/pdctl/ctl_test.go:
```diff
@@ -0,0 +1,70 @@
+	return cmd
+}
+
+func Test_genCompleter(t *testing.T) {
```
```suggestion
func TestGenCompleter(t *testing.T) {
```

jiashiwen replied:
ok

rleungx commented on tools/pd-ctl/pdctl/ctl_test.go:
```diff
@@ -0,0 +1,70 @@
+	pc := genCompleter(rootCmd)
+
+	for _, cmd := range subcommand {
+		runarray := []rune(cmd)
+		inprefixarray := true
```
```suggestion
		runArray := []rune(cmd)
		inPrefixArray := true
```

jiashiwen replied:
all right

rleungx reviewed (commented):
The rest LGTM

howardlau1999 commented on tools/pd-ctl/pdctl/ctl.go:
```diff
@@ -190,3 +194,23 @@ func loop() {
+
+func genCompleter(cmd *cobra.Command) []readline.PrefixCompleterInterface {
+	pc := []readline.PrefixCompleterInterface{}
+	if len(cmd.Commands()) != 0 {
```
is this necessary?

jiashiwen replied:
ok,remove it

howardlau1999 commented on tools/pd-ctl/pdctl/ctl_test.go:
```diff
@@ -0,0 +1,70 @@
+			}
+		}
+
+		if inPrefixArray == false {
+			t.Errorf("%s not in prefix array", cmd)
+		}
```
what is this log for? why not use assert?

howardlau1999 commented:
LGTM

jiashiwen commented:
`for _, v := range cmd.Commands() {
		if v.HasFlags() {
			flagsPc := []readline.PrefixCompleterInterface{}
			flagUsages := strings.Split(strings.Trim(v.Flags().FlagUsages(), " "), "\n")
			for i := 0; i < len(flagUsages)-1; i++ {
				flagsPc = append(flagsPc, readline.PcItem(strings.Split(strings.Trim(flagUsages[i], " "), " ")[0]))
			}
			flagsPc = append(flagsPc, genCompleter(v)...)
			pc = append(pc, readline.PcItem(strings.Split(v.Use, " ")[0], flagsPc...))
		} else {`

nolouch reviewed (commented):
LGTM.

nolouch commented:
/merge
